### PR TITLE
Further parallelize linspace in addition to AVX

### PR DESCRIPTION
**Pull request by xuhdev**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38093 Further parallelize linspace in addition to AVX**
* #37981 Allow specifying range in and cpu_serial_kernel and cpu_serial_kernel_vec

Benchmark (gcc 8.3, Debian Buster, turbo off, Release build, Intel(R) Xeon(R) E-2136, Parallelization using OpenMP):

```
import timeit
for dtype in ('torch.double', 'torch.float', 'torch.uint8', 'torch.int8', 'torch.int16', 'torch.int32', 'torch.int64'):
    for n, t in [(40_000, 50000),
                (400_000, 5000)]:
        print(f'torch.linspace(0, 10, {n}, dtype={dtype}) for {t} times')
        print(timeit.timeit(f'torch.linspace(0, 10, {n}, dtype={dtype})', setup=f'import torch', number=t))
```

With AVX
========

Before:

```
torch.linspace(0, 10, 40000, dtype=torch.double) for 50000 times
1.0942596640015836
torch.linspace(0, 10, 400000, dtype=torch.double) for 5000 times
0.9209065200011537
torch.linspace(0, 10, 40000, dtype=torch.float) for 50000 times
1.0520610109997506
torch.linspace(0, 10, 400000, dtype=torch.float) for 5000 times
0.9031864690005023
torch.linspace(0, 10, 40000, dtype=torch.uint8) for 50000 times
1.949299545998656
torch.linspace(0, 10, 400000, dtype=torch.uint8) for 5000 times
1.82629113800067
torch.linspace(0, 10, 40000, dtype=torch.int8) for 50000 times
1.9547776939980395
torch.linspace(0, 10, 400000, dtype=torch.int8) for 5000 times
1.8259895039991534
torch.linspace(0, 10, 40000, dtype=torch.int16) for 50000 times
2.759497356000793
torch.linspace(0, 10, 400000, dtype=torch.int16) for 5000 times
2.6285490109985403
torch.linspace(0, 10, 40000, dtype=torch.int32) for 50000 times
2.3456633150017296
torch.linspace(0, 10, 400000, dtype=torch.int32) for 5000 times
2.2031515989983745
torch.linspace(0, 10, 40000, dtype=torch.int64) for 50000 times
2.559069258000818
torch.linspace(0, 10, 400000, dtype=torch.int64) for 5000 times
2.378239962999942
```

After:

```
torch.linspace(0, 10, 40000, dtype=torch.double) for 50000 times
0.8100852870011295
torch.linspace(0, 10, 400000, dtype=torch.double) for 5000 times
0.18943897200006177
torch.linspace(0, 10, 40000, dtype=torch.float) for 50000 times
0.6679975400002149
torch.linspace(0, 10, 400000, dtype=torch.float) for 5000 times
0.17846923400065862
torch.linspace(0, 10, 40000, dtype=torch.uint8) for 50000 times
1.1431112539976311
torch.linspace(0, 10, 400000, dtype=torch.uint8) for 5000 times
0.3336703610002587
torch.linspace(0, 10, 40000, dtype=torch.int8) for 50000 times
1.157699686998967
torch.linspace(0, 10, 400000, dtype=torch.int8) for 5000 times
0.32964968899977976
torch.linspace(0, 10, 40000, dtype=torch.int16) for 50000 times
1.5379577429994242
torch.linspace(0, 10, 400000, dtype=torch.int16) for 5000 times
0.4638638729993545
torch.linspace(0, 10, 40000, dtype=torch.int32) for 50000 times
1.360489848000725
torch.linspace(0, 10, 400000, dtype=torch.int32) for 5000 times
0.4033017760011717
torch.linspace(0, 10, 40000, dtype=torch.int64) for 50000 times
1.4591587399991113
torch.linspace(0, 10, 400000, dtype=torch.int64) for 5000 times
0.44132660000104806
```

Without AVX
===========

Before:

```
torch.linspace(0, 10, 40000, dtype=torch.double) for 50000 times
3.4967273879992717
torch.linspace(0, 10, 400000, dtype=torch.double) for 5000 times
3.330881046000286
torch.linspace(0, 10, 40000, dtype=torch.float) for 50000 times
2.176502857997548
torch.linspace(0, 10, 400000, dtype=torch.float) for 5000 times
2.023505228000431
torch.linspace(0, 10, 40000, dtype=torch.uint8) for 50000 times
2.117801246000454
torch.linspace(0, 10, 400000, dtype=torch.uint8) for 5000 times
1.9885458380013006
torch.linspace(0, 10, 40000, dtype=torch.int8) for 50000 times
2.1057261179994384
torch.linspace(0, 10, 400000, dtype=torch.int8) for 5000 times
1.9809251260012388
torch.linspace(0, 10, 40000, dtype=torch.int16) for 50000 times
3.187070896001387
torch.linspace(0, 10, 400000, dtype=torch.int16) for 5000 times
3.049615387000813
torch.linspace(0, 10, 40000, dtype=torch.int32) for 50000 times
3.4874590049985272
torch.linspace(0, 10, 400000, dtype=torch.int32) for 5000 times
3.33596555099939
torch.linspace(0, 10, 40000, dtype=torch.int64) for 50000 times
4.256659758000751
torch.linspace(0, 10, 400000, dtype=torch.int64) for 5000 times
4.100936053000623
```

After:

```
torch.linspace(0, 10, 40000, dtype=torch.double) for 50000 times
1.9155298300029244
torch.linspace(0, 10, 400000, dtype=torch.double) for 5000 times
0.598213522000151
torch.linspace(0, 10, 40000, dtype=torch.float) for 50000 times
1.3183841649988608
torch.linspace(0, 10, 400000, dtype=torch.float) for 5000 times
0.40136947100108955
torch.linspace(0, 10, 40000, dtype=torch.uint8) for 50000 times
1.2191377319977619
torch.linspace(0, 10, 400000, dtype=torch.uint8) for 5000 times
0.35984685299990815
torch.linspace(0, 10, 40000, dtype=torch.int8) for 50000 times
1.2153874989999167
torch.linspace(0, 10, 400000, dtype=torch.int8) for 5000 times
0.35752785600197967
torch.linspace(0, 10, 40000, dtype=torch.int16) for 50000 times
1.750796647000243
torch.linspace(0, 10, 400000, dtype=torch.int16) for 5000 times
0.5376063230032742
torch.linspace(0, 10, 40000, dtype=torch.int32) for 50000 times
1.9153429929974664
torch.linspace(0, 10, 400000, dtype=torch.int32) for 5000 times
0.5952553579991218
torch.linspace(0, 10, 40000, dtype=torch.int64) for 50000 times
2.281823589000851
torch.linspace(0, 10, 400000, dtype=torch.int64) for 5000 times
0.7391443560009066
```

Differential Revision: [D21528099](https://our.internmc.facebook.com/intern/diff/D21528099)

### CODE REVIEW

**Comment by xuhdev** [comment]
If this works, then we should be able to vectorize most other range factories similarly to what has been done here.

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 1]
```diff
@@ -13,33 +13,37 @@ namespace {
 using namespace vec256;
 
 static void linspace_kernel(TensorIterator& iter, Scalar scalar_start, Scalar scalar_end, int64_t steps) {
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "linspace_cpu", [&]() {
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "linspace_cpu", [=]() {
```
Comment (malfet) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:16]: Why this is needed? Capture by reference should always be faster, shouldn't it?

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 2]
```diff
@@ -13,33 +13,37 @@ namespace {
...
-          Vec256<scalar_t> res;
-          if (idx < halfway) {
-            res = Vec256<scalar_t>::arange(start + step * idx, step);
-          } else {
-            res = Vec256<scalar_t>::arange(end - step * (steps - idx - 1), step);
-          }
-          idx += Vec256<scalar_t>::size();
-          return res;
-        });
+    at::parallel_for(0, steps, internal::GRAIN_SIZE, [=](int64_t p_begin, int64_t p_end) {
```
Comment (malfet) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:23]: Again, why capture by value?

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp (hunk 2 above)
Comment (xuhdev) [reply aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp]: No special reason. I changed to reference now. A previous version wants to have a new copy of `iter` via the capture, but it turned out not working.

**Review by malfet — CHANGES_REQUESTED** [review]
@xuhdev please rebase this PR, as it can not be cleanly merged into master.

**Comment by xuhdev** [comment]
rebased

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 3]
```diff
@@ -16,30 +16,34 @@ static void linspace_kernel(TensorIterator& iter, Scalar scalar_start, Scalar sc
...
-          return res;
-        });
+    at::parallel_for(0, steps, internal::GRAIN_SIZE, [&](int64_t p_begin, int64_t p_end) {
+      int64_t idx(p_begin);
+      TensorIterator it(iter);
+      cpu_serial_kernel_vec(
+          it,
+          [start, end, step, halfway, steps, &idx]() -> scalar_t {
+            if (idx < halfway) {
+              return start + step * (idx++);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:30]: Please apply this one
```suggestion
              return start + step * static_cast<step_t>(idx++);
```

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 4]
```diff
@@ -16,30 +16,34 @@ static void linspace_kernel(TensorIterator& iter, Scalar scalar_start, Scalar sc
...
+            } else {
+              return end - step * (steps - (idx++) - 1);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:32]:
```suggestion
              return end - step * static_cast<step_t>(steps - (idx++) - 1);
```

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 5]
```diff
@@ -16,30 +16,34 @@ static void linspace_kernel(TensorIterator& iter, Scalar scalar_start, Scalar sc
...
+            }
+          },
+          [start, end, step, halfway, steps, &idx]() -> Vec256<scalar_t> {
+            Vec256<scalar_t> res;
+            if (idx < halfway) {
+              res = Vec256<scalar_t>::arange(start + step * idx, step);
+            } else {
+              res = Vec256<scalar_t>::arange(
+                  end - step * (steps - idx - 1), step);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:41]:
```suggestion
                  end - step * static_cast<step_t>(steps - idx - 1), step);
```

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 6]
```diff
@@ -16,30 +16,34 @@ static void linspace_kernel(TensorIterator& iter, Scalar scalar_start, Scalar sc
...
+            if (idx < halfway) {
+              return start + step * (idx++);
+            } else {
+              return end - step * (steps - (idx++) - 1);
+            }
+          },
+          [start, end, step, halfway, steps, &idx]() -> Vec256<scalar_t> {
+            Vec256<scalar_t> res;
+            if (idx < halfway) {
+              res = Vec256<scalar_t>::arange(start + step * idx, step);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:38]:
```suggestion
              res = Vec256<scalar_t>::arange(start + step * static_cast<step_t>(idx), step);
```

**Review by malfet — COMMENTED** [review]
Looks like you've missed a few changes during the rebase

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp [hunk 7]
```diff
@@ -16,30 +16,34 @@ static void linspace_kernel(TensorIterator& iter, Scalar scalar_start, Scalar sc
   AT_DISPATCH_ALL_TYPES_AND_C10_COMPLEX(iter.dtype(), "linspace_cpu", [&]() {
     // step should be of double type for all integral types
     using step_t = std::conditional_t<std::is_integral<scalar_t>::value, double, scalar_t>;
-    int64_t idx(0);
-    scalar_t start = scalar_start.to<scalar_t>();
-    scalar_t end = scalar_end.to<scalar_t>();
-    step_t step = static_cast<step_t>(end - start) / static_cast<step_t>(steps - 1);
+    const scalar_t start = scalar_start.to<scalar_t>();
+    const scalar_t end = scalar_end.to<scalar_t>();
+    const step_t step = static_cast<step_t>(end - start) / (steps - 1);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp:21]:
```suggestion
    const step_t step = static_cast<step_t>(end - start) / static_cast<step_t>(steps - 1);
```

**Review by malfet — APPROVED** [review]
Will land if CI is green

File: aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp (hunk 7 above)
Comment (xuhdev) [reply aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp]: Thanks for catching this! Interesting, I was confused by this comment and thought they have removed the cast: https://github.com/pytorch/pytorch/pull/38021#discussion_r423658548

We can deal with the cast in another PR though.

**Pull request closed without merging**